File: Session 11/Hamiltonian_Cycle.cpp
#include<iostream>
#include<vector>
#define fastio ios::sync_with_stdio(false); cin.tie(nullptr);
using namespace std;
vector<vector<int>>adjList;
vector<int>path;
vector<bool>vis;
int v=5;
bool hasEdge(int u,int v){
    for(int x:adjList[u]){
        if(x==v){
            return true;
        }
    }
    return false;
}
bool solve(int u,int count){
    if(count==v){
        return hasEdge(u,path[0]);
    }
    for(int v:adjList[u]){
        if(!vis[v]){
            vis[v]=true;
            path.push_back(v);
            if(solve(v,count+1)){
                return true;
            }
            vis[v]=false;
            path.pop_back();
        }
    }
    return false;
}

int main(){
    fastio;
    int t=1;
    while(t--){
        adjList={
            {1,3},
            {0,2,3,4},
            {1,4},
            {0,1,4},
            {1,2,3}
        };
        vis.assign(v,false);
        path.push_back(0);
        vis[0]=true;
        if (solve(0, 1)){
        cout<<"Hamiltonian Cycle exists:\n";
        for (int x :path){
        cout<<x<< " ";
        }
        cout<<path[0]<<"\n";
        } 
        else{
        cout<<"No Hamiltonian Cycle\n";
        }
    }
    return 0;
}

File: Assignment/Hamiltonian_Cycle.java
public class HamiltonianCycle {
    private int V;
    private int[] path;
    public HamiltonianCycle(int v) {
        this.V = v;
        path = new int[V];
    }
    public boolean solve(int[][] graph) {
        for (int i = 0; i < V; i++) {
            path[i] = -1;
        }
        path[0] = 0;
        if (!hamCycleUtil(graph, 1)) {
            System.out.println("No Hamiltonian Cycle exists");
            return false;
        }
        printSolution();
        return true;
    }
    private boolean hamCycleUtil(int[][] graph, int pos) {
        if (pos == V) {
            return graph[path[pos - 1]][path[0]] == 1;
        }
        for (int v = 1; v < V; v++) {
            if (isSafe(v, graph, pos)) {
                path[pos] = v;
                if (hamCycleUtil(graph, pos + 1)) {
                    return true;
                }
                path[pos] = -1;
            }
        }
        return false;
    }
    private boolean isSafe(int v, int[][] graph, int pos) {
        if (graph[path[pos - 1]][v] == 0) return false;

        for (int i = 0; i < pos; i++) {
            if (path[i] == v) return false;
        }
        return true;
    }
    private void printSolution() {
        System.out.print("Hamiltonian Cycle: ");
        for (int i = 0; i < V; i++) {
            System.out.print(path[i] + " ");
        }
        System.out.println(path[0]);
    }
    public static void main(String[] args) {
        int[][] graph = {
            {0, 1, 1, 0, 1},
            {1, 0, 1, 1, 1},
            {1, 1, 0, 1, 0},
            {0, 1, 1, 0, 1},
            {1, 1, 0, 1, 0}
        };
        HamiltonianCycle hc = new HamiltonianCycle(graph.length);
        hc.solve(graph);
    }
}
